core/primitives/src/validator_mandates.rs

```diff
@@ -0,0 +1,259 @@
+            validators.iter().map(|v| v.num_mandates(config.stake_per_mandate)).collect();
+        let num_total_mandates =
+            num_mandates_per_validator.iter().map(|&num| usize::from(num)).sum();
+        let mut mandates: Vec<ValidatorId> = Vec::with_capacity(num_total_mandates);
+
+        for i in 0..validators.len() {
+            for _ in 0..num_mandates_per_validator[i] {
+                // Each validator's position corresponds to its id.
+                mandates.push(i as ValidatorId);
+            }
+        }
+
+        let required_mandates = config.min_mandates_per_shard * config.num_shards;
+        if mandates.len() < required_mandates {
+            // TODO(chunk-validator-assignment) dynamically lower `stake_per_mandate` to reach enough mandates
+            panic!(
+                "not enough validator mandates: got {}, need {}",
+                mandates.len(),
+                required_mandates
+            );
+        }
+
+        Self { config, mandates }
+    }
+
+    /// Returns a validator assignment obtained by shuffling mandates.
+    ///
+    /// It clones mandates since [`ValidatorMandates`] is supposed to be valid for an epoch, while a
+    /// new assignment is calculated at every height.
+    ///
+    /// # Interpretation of the return value
+    ///
+    /// The returned vector contains one map per shard, with the position in the vector
+    /// corresponding to `shard_id` in `0..num_shards`.
+    ///
+    /// Each `HashMap` maps `ValidatorId`s to the number of mandates they have in the corresponding
+    /// shards. A validator whose id is not in a map has not been assigned to the shard.
+    ///
+    /// ## Example
+    ///
+    /// Let `res` be the return value of this function, then `res[0][1]` maps to the number of
+    /// mandates validator with `ValidatorId` 1 holds in shard with id 0.
+    pub fn sample<R>(&self, rng: &mut R) -> Vec<HashMap<ValidatorId, u16>>
+    where
+        R: Rng + ?Sized,
+    {
+        let shuffled_mandates = self.shuffle(rng);
+
+        // Assign shuffled seat at position `i` to the shard with id `i % num_shards`.
+        let mut assignments_per_shard = Vec::with_capacity(self.config.num_shards);
+        for shard_id in 0..self.config.num_shards {
+            let mut assignments = HashMap::new();
+            let mut idx = shard_id;
+
+            while idx < shuffled_mandates.len() {
+                let id = shuffled_mandates[idx];
+                assignments.entry(id).and_modify(|counter| *counter += 1).or_insert(1);
+                idx += self.config.num_shards;
+            }
+
+            assignments_per_shard.push(assignments)
+        }
+
+        assignments_per_shard
+    }
+
+    /// Clones the contained mandates and shuffles them. Cloning is required as a shuffle happens at
+    /// every height while the `ValidatorMandates` are to be valid for an epoch.
+    fn shuffle<R>(&self, rng: &mut R) -> Vec<ValidatorId>
```

Comment:
nit: might be clearer to call it "shuffled"

Reply:
Renamed in 8ab0212.